File: python/dsa/data-structures/stack-queue/next-greater-element-i.py
def nextGreaterElement(n1: List[int], n2: List[int]) -> List[int]:
    nxtGreaterMap = {} # @viz: init_map
    decreStack = [] # @viz: init_stack
    
    for x in n2: # @viz: loop
        while decreStack and x > decreStack[-1]: # @viz: while_check
            nxtGreaterMap[decreStack.pop()] = x # @viz: map_set
        decreStack.append(x) # @viz: push
        
    return [nxtGreaterMap.get(n, -1) for n in n1] # @viz: finish

File: cpp/dsa/data-structures/stack-queue/next-greater-element-i.cpp
class Solution {
public:
    vector<int> nextGreaterElement(vector<int>& n1, vector<int>& n2) {
        unordered_map<int, int> nMap; // @viz: init_map
        stack<int> s; // @viz: init_stack
        
        for (int val : n2) { // @viz: loop
            while (!s.empty() && val > s.top()) { // @viz: while_check
                nMap[s.top()] = val; // @viz: map_set
                s.pop();
            }
            s.push(val); // @viz: push
        }
        
        vector<int> result;
        for (int x : n1) {
            result.push_back(nMap.count(x) ? nMap[x] : -1);
        }
        return result; // @viz: finish
    }
}